### PR TITLE
Return typed errors for Azure client interfaces

### PR DESCRIPTION
**Pull request by feiskyer**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

/kind feature
/sig cloud-provider
/area provider/azure
/priority important-soon

**What this PR does / why we need it**:

Continue of https://github.com/kubernetes-sigs/cloud-provider-azure/issues/247, return typed errors for Azure client interfaces:

* Add a new `retry.Error`
* Return `retry.Error` for Azure client interfaces
* Switch to new returned errors in other places

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/azure/retry/azure_error.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+// Error indicates an error returned by Azure APIs.
+type Error struct {
+	// Retriable indicates whether the request is retriable.
+	Retriable bool
+	// HTTPStatusCode indicates the response HTTP status code.
+	HTTPStatusCode int
+	// RetryAfter indicates the time when the request should retry after throttling.
+	// A throttled request is retriable.
+	RetryAfter time.Time
```
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/retry/azure_error.go]: `RetryAfter` is not used in this PR, right?

Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/retry/azure_error.go]: Right, it would be used in future PRs.

**Review by andyzhangx — APPROVED** [review]
/lgtm
/test pull-kubernetes-e2e-aks-engine-azure
/test pull-kubernetes-e2e-azure-disk
/test pull-kubernetes-e2e-azure-disk-vmss
/hold
hold until all tests pass

**Comment by feiskyer** [comment]
test passed.
/hold cancel

**Comment by feiskyer** [comment]
@dims @liggitt could you help to approve the vendor updates?

/assign @dims @liggitt

**Comment by liggitt** [comment]
/approve
for modules.txt change 
all other packages already approved by existing owners

**Pull request merged**